File: src/shortcuts.ts
import type { MainView, RightPanelKind } from "./types";
import { readAppSettings } from "./settings";
import { tm } from "./i18n";
import { isAppleShortcutPlatform } from "./platform";

export type ShortcutScope = "global" | "workspace" | "project-sidebar" | "task-sidebar" | "right-sidebar";

export type ShortcutContext = {
  focusScope: ShortcutScope;
  mainView: MainView;
  rightPanel: RightPanelKind | null;
};

export type ShortcutHandler = (event: KeyboardEvent, context: ShortcutContext) => boolean;

export type ShortcutDefinition = {
  id: string;
  label: string;
  labelKey: string;
  keys: Array<{
    key: string;
    meta?: boolean;
    ctrl?: boolean;
    alt?: boolean;
    shift?: boolean;
  }>;
  scope: ShortcutScope;
};

const commonShortcutKeys: Record<string, ShortcutDefinition["keys"]> = {
  "terminal.split": [{ key: "\\", meta: true, shift: true }],
  "terminal.tab": [{ key: "t", meta: true, shift: true }],
  "panel.git": [{ key: "g", meta: true, shift: true }],
  "panel.ai": [{ key: "a", meta: true, shift: true }],
};

const terminalCommandsToSkipShell = new Set([
  "close.active",
  "panel.ai",
  "panel.git",
  "terminal.split",
  "terminal.tab",
  "view.files",
  "view.review",
  "view.terminal",
]);

type ShortcutRegistration = {
  id: number;
  scope: ShortcutScope;
  handler: ShortcutHandler;
};

export const appShortcutDefinitions: ShortcutDefinition[] = [
  {
    id: "view.terminal",
    label: "Terminal View",
    labelKey: "shortcut.view.terminal",
    scope: "global",
    keys: [{ key: "1", meta: true }],
  },
  {
    id: "view.files",
    label: "Files View",
    labelKey: "shortcut.view.files",
    scope: "global",
    keys: [{ key: "2", meta: true }],
  },
  {
    id: "view.review",
    label: "Review View",
    labelKey: "shortcut.view.review",
    scope: "global",
    keys: [{ key: "3", meta: true }],
  },
  {
    id: "project.create",
    label: "New Project",
    labelKey: "shortcut.project.create",
    scope: "project-sidebar",
    keys: [{ key: "n", meta: true }],
  },
  {
    id: "settings.open",
    label: "Open Settings",
    labelKey: "shortcut.settings.open",
    scope: "project-sidebar",
    keys: [{ key: ",", meta: true }],
  },
  {
    id: "task.create",
    label: "New Worktree",
    labelKey: "shortcut.task.create",
    scope: "task-sidebar",
    keys: [{ key: "n", meta: true }],
  },
  {
    id: "editor.save",
    label: "Save File",
    labelKey: "shortcut.editor.save",
    scope: "workspace",
    keys: [{ key: "s", meta: true }],
  },
  {
    id: "editor.search",
    label: "Search Files",
    labelKey: "shortcut.editor.search",
    scope: "workspace",
    keys: [{ key: "f", meta: true }],
  },
  {
    id: "close.active",
    label: "Close Current Project",
    labelKey: "shortcut.close.active",
    scope: "workspace",
    keys: [{ key: "w", meta: true }],
  },
];

const registrations: ShortcutRegistration[] = [];
let nextId = 1;

export function registerShortcutHandler(scope: ShortcutScope, handler: ShortcutHandler) {
  const registration = {
    id: nextId,
    scope,
    handler,
  };
  nextId += 1;
  registrations.push(registration);

  return () => {
    const index = registrations.findIndex((item) => item.id === registration.id);
    if (index >= 0) {
      registrations.splice(index, 1);
    }
  };
}

export function dispatchShortcut(event: KeyboardEvent, context: ShortcutContext) {
  const scopes = context.focusScope === "global" ? ["global"] : [context.focusScope, "global"];

  for (const scope of scopes) {
    const handlers = registrations.filter((registration) => registration.scope === scope).sort((a, b) => b.id - a.id);

    for (const registration of handlers) {
      if (registration.handler(event, context)) {
        event.preventDefault();
        event.stopPropagation();
        return true;
      }
    }
  }

  return false;
}

export function configuredShortcutKeys(id: string) {
  const configured = parseShortcutSequence(readAppSettings().shortcuts[id]);
  if (configured.length > 0) return configured;
  return appShortcutDefinitions.find((shortcut) => shortcut.id === id)?.keys ?? commonShortcutKeys[id] ?? [];
}

export function isConfiguredShortcut(event: KeyboardEvent, id: string) {
  return configuredShortcutKeys(id).some((shortcut) => isShortcut(event, shortcut));
}

export function shouldTerminalSkipShell(event: KeyboardEvent) {
  return [...terminalCommandsToSkipShell].some((id) => isConfiguredShortcut(event, id));
}

export function shortcutDisplayValue(id: string, fallback?: string) {
  const keys = configuredShortcutKeys(id);
  return keys.length > 0 ? keys.map(formatShortcutKeys).join(" / ") : (fallback ?? tm("settings.shortcut.unset", "Not Set"));
}

export function isShortcut(
  event: KeyboardEvent,
  shortcut: {
    key: string;
    meta?: boolean;
    ctrl?: boolean;
    alt?: boolean;
    shift?: boolean;
  },
) {
  const usesApplePrimary = isAppleShortcutPlatform();
  const expectsPrimary = Boolean(shortcut.meta);
  const expectedMeta = usesApplePrimary ? expectsPrimary : false;
  const expectedCtrl = Boolean(shortcut.ctrl) || (!usesApplePrimary && expectsPrimary);
  return (
    event.key.toLowerCase() === shortcut.key.toLowerCase() &&
    event.metaKey === expectedMeta &&
    event.ctrlKey === expectedCtrl &&
    event.altKey === Boolean(shortcut.alt) &&
    event.shiftKey === Boolean(shortcut.shift)
  );
}

export function formatShortcutKeys(shortcut: ShortcutDefinition["keys"][number]) {
  const usesAppleSymbols = isAppleShortcutPlatform();
  const parts = [
    shortcut.meta ? (usesAppleSymbols ? "⌘" : "Ctrl+") : "",
    shortcut.ctrl ? (usesAppleSymbols ? "⌃" : "Ctrl+") : "",
    shortcut.alt ? (usesAppleSymbols ? "⌥" : "Alt+") : "",
    shortcut.shift ? (usesAppleSymbols ? "⇧" : "Shift+") : "",
    shortcut.key.length === 1 ? shortcut.key.toUpperCase() : shortcut.key,
  ];
  return parts.filter(Boolean).join("");
}

export function parseShortcutSequence(value?: string) {
  const raw = value?.trim();
  if (!raw) return [];
  return raw
    .split("/")
    .map((item) => parseShortcutText(item.trim()))
    .filter((item): item is ShortcutDefinition["keys"][number] => Boolean(item));
}

export function serializeShortcutSequence(keys: ShortcutDefinition["keys"]) {
  return keys.map(formatShortcutKeys).join(" / ");
}

function parseShortcutText(value: string): ShortcutDefinition["keys"][number] | null {
  if (!value) return null;
  let rest = value
    .replace(/Command|Cmd|⌘/gi, "Meta+")
    .replace(/Control|Ctrl|⌃/gi, "Ctrl+")
    .replace(/Option|Alt|⌥/gi, "Alt+")
    .replace(/Shift|⇧/gi, "Shift+")
    .replace(/\s+/g, "");
  const meta = rest.includes("Meta+");
  const ctrl = rest.includes("Ctrl+");
  const alt = rest.includes("Alt+");
  const shift = rest.includes("Shift+");
  rest = rest.replace(/Meta\+|Ctrl\+|Alt\+|Shift\+/g, "");
  if (!rest) return null;
  return {
    key: rest.length === 1 ? rest.toLowerCase() : rest,
    meta,
    ctrl,
    alt,
    shift,
  };
}

File: apps/desktop/src/app/shortcuts.rs
use std::collections::HashMap;

pub(in crate::app) fn shortcut_display_from_keystroke(keystroke: &gpui::Keystroke) -> String {
    let mut parts = Vec::new();
    if keystroke.modifiers.platform {
        parts.push(
            if cfg!(target_os = "macos") {
                "⌘"
            } else {
                "Ctrl+"
            }
            .to_string(),
        );
    }
    if keystroke.modifiers.control {
        parts.push(
            if cfg!(target_os = "macos") {
                "⌃"
            } else {
                "Ctrl+"
            }
            .to_string(),
        );
    }
    if keystroke.modifiers.alt {
        parts.push(
            if cfg!(target_os = "macos") {
                "⌥"
            } else {
                "Alt+"
            }
            .to_string(),
        );
    }
    if keystroke.modifiers.shift {
        parts.push(
            if cfg!(target_os = "macos") {
                "⇧"
            } else {
                "Shift+"
            }
            .to_string(),
        );
    }
    let key = if keystroke.key.chars().count() == 1 {
        keystroke.key.to_uppercase()
    } else {
        keystroke.key.clone()
    };
    parts.push(key);
    parts.join("")
}

fn default_shortcut_display(shortcut_id: &str) -> Option<&'static str> {
    let primary = if cfg!(target_os = "macos") {
        "⌘"
    } else {
        "Ctrl+"
    };
    match (shortcut_id, primary) {
        ("view.terminal", "⌘") => Some("⌘⌥1"),
        ("view.files", "⌘") => Some("⌘⌥2"),
        ("view.review", "⌘") => Some("⌘⌥3"),
        ("project.create", "⌘") => Some("⌘N"),
        ("project.open_folder", "⌘") => Some("⌘O"),
        ("settings.open", "⌘") => Some("⌘,"),
        ("task.create", "⌘") => Some("⌘⇧N"),
        ("editor.save", "⌘") => Some("⌘S"),
        ("editor.search", "⌘") => Some("⌘F"),
        ("close.active", "⌘") => Some("⌘W"),
        ("sidebar.projects.toggle", "⌘") => Some("⌘⌥P"),
        ("sidebar.tasks.toggle", "⌘") => Some("⌘⌥T"),
        ("assistant.git.open", "⌘") => Some("⌘⇧G"),
        ("panel.git", "⌘") => Some("⌘⇧G"),
        ("assistant.files.open", "⌘") => Some("⌘⇧F"),
        ("assistant.ai.open", "⌘") => Some("⌘⇧A"),
        ("panel.ai", "⌘") => Some("⌘⇧A"),
        ("assistant.ssh.open", "⌘") => Some("⌘⇧S"),
        ("terminal.split", "⌘") => Some("⌘D"),
        ("terminal.split.create", "⌘") => Some("⌘D"),
        ("terminal.split.vertical", "⌘") => Some("⌘⇧D"),
        ("terminal.split.close", "⌘") => Some("⌃D"),
        ("view.terminal", _) => Some("Ctrl+Alt+1"),
        ("view.files", _) => Some("Ctrl+Alt+2"),
        ("view.review", _) => Some("Ctrl+Alt+3"),
        ("project.create", _) => Some("Ctrl+N"),
        ("project.open_folder", _) => Some("Ctrl+O"),
        ("settings.open", _) => Some("Ctrl+,"),
        ("task.create", _) => Some("Ctrl+Shift+N"),
        ("editor.save", _) => Some("Ctrl+S"),
        ("editor.search", _) => Some("Ctrl+F"),
        ("close.active", _) => Some("Ctrl+W"),
        ("sidebar.projects.toggle", _) => Some("Ctrl+Alt+P"),
        ("sidebar.tasks.toggle", _) => Some("Ctrl+Alt+T"),
        ("assistant.git.open", _) => Some("Ctrl+Shift+G"),
        ("panel.git", _) => Some("Ctrl+Shift+G"),
        ("assistant.files.open", _) => Some("Ctrl+Shift+F"),
        ("assistant.ai.open", _) => Some("Ctrl+Shift+A"),
        ("panel.ai", _) => Some("Ctrl+Shift+A"),
        ("assistant.ssh.open", _) => Some("Ctrl+Shift+S"),
        ("terminal.split", _) => Some("Ctrl+D"),
        ("terminal.split.create", _) => Some("Ctrl+D"),
        ("terminal.split.vertical", _) => Some("Ctrl+Shift+D"),
        ("terminal.split.close", _) => Some("Ctrl+Alt+D"),
        _ => None,
    }
}

pub(in crate::app) fn normalized_shortcut_text(value: &str) -> Option<String> {
    let mut rest = value.trim().to_lowercase();
    if rest.is_empty() {
        return None;
    }

    let platform = rest.contains("command") || rest.contains("cmd") || rest.contains('⌘');
    let control = rest.contains("control") || rest.contains("ctrl") || rest.contains('⌃');
    let alt = rest.contains("option") || rest.contains("alt") || rest.contains('⌥');
    let shift = rest.contains("shift") || rest.contains('⇧');

    for token in [
        "command", "cmd", "control", "ctrl", "option", "alt", "shift", "⌘", "⌃", "⌥", "⇧", "+",
    ] {
        rest = rest.replace(token, "");
    }
    rest.retain(|character| !character.is_whitespace());
    if rest.is_empty() {
        return None;
    }

    let key = if rest.chars().count() == 1 {
        rest.to_uppercase()
    } else {
        rest
    };
    Some(format!(
        "{}{}{}{}{}",
        if platform { "Meta+" } else { "" },
        if control { "Ctrl+" } else { "" },
        if alt { "Alt+" } else { "" },
        if shift { "Shift+" } else { "" },
        key
    ))
}

fn shortcut_value_matches(configured: &str, actual: &str) -> bool {
    let Some(actual) = normalized_shortcut_text(actual) else {
        return false;
    };
    configured
        .split('/')
        .filter_map(normalized_shortcut_text)
        .any(|candidate| candidate == actual)
}

pub(in crate::app) fn shortcut_matches(
    shortcuts: &HashMap<String, String>,
    shortcut_id: &str,
    actual: &str,
) -> bool {
    shortcuts
        .get(shortcut_id)
        .filter(|value| !value.trim().is_empty())
        .map(|value| shortcut_value_matches(value, actual))
        .unwrap_or_else(|| {
            default_shortcut_display(shortcut_id)
                .map(|value| shortcut_value_matches(value, actual))
                .unwrap_or(false)
        })
}
